File: services/ingestion/src/domain/events.test.ts
import { describe, expect, it } from "vitest";

import { buildEventId, normalizeChainEvent, type ParsedChainEvent } from "./events.js";

describe("normalizeChainEvent", () => {
  it("adds canonical envelope for VoteCommitted", () => {
    const parsed: ParsedChainEvent = {
      event_type: "VoteCommitted",
      payload: {
        proposal_id: "p1",
        voter_pubkey: "voter1",
        commitment: "2fHb8QiezB2CSfXhwtZ9WaJ81HCtGJhP5eXEbQCwcbuz",
      },
      slot: 42,
      tx_signature: "sig-abc",
    };

    const event = normalizeChainEvent(parsed, 0, "voting-contract", 1, 1_700_000_000);

    expect(event.event_id).toBe(buildEventId("sig-abc", "VoteCommitted", 0));
    expect(event.event_type).toBe("VoteCommitted");
    expect(event.proposal_id).toBe("p1");
    expect(event.commitment).toBe("2fHb8QiezB2CSfXhwtZ9WaJ81HCtGJhP5eXEbQCwcbuz");
    expect(event.slot).toBe(42);
  });

  it("adds canonical envelope for VoteRevealed", () => {
    const parsed: ParsedChainEvent = {
      event_type: "VoteRevealed",
      payload: {
        proposal_id: "p1",
        option_id: "yes",
        voter_pubkey: "voter1",
      },
      slot: 1,
      tx_signature: "sig",
    };

    const event = normalizeChainEvent(parsed, 0, "voting-contract", 1, 1);
    expect(event.event_type).toBe("VoteRevealed");
    expect(event.option_id).toBe("yes");
  });

  it("requires at least two options for ProposalCreated", () => {
    const parsed: ParsedChainEvent = {
      event_type: "ProposalCreated",
      payload: {
        proposal_id: "p1",
        title: "Test",
        options: ["only"],
        commit_ends_at: 1,
        reveal_ends_at: 2,
        phase: "commit",
      },
      slot: 1,
      tx_signature: "sig",
    };

    expect(() => normalizeChainEvent(parsed, 0, "voting-contract", 1, 1)).toThrow("at least two");
  });
});

File: services/aggregator/internal/app/domain/events_test.go
package domain_test

import (
	"errors"
	"testing"

	"github.com/chain-to-cloud/aggregator/internal/app/domain"
)

func validVoteRevealedJSON() []byte {
	return []byte(`{
		"event_id": "sig:VoteRevealed:0",
		"event_type": "VoteRevealed",
		"timestamp": 1700000000,
		"source": "voting-contract",
		"version": 1,
		"proposal_id": "p1",
		"option_id": "1",
		"voter_pubkey": "voter1",
		"slot": 1,
		"tx_signature": "sig"
	}`)
}

func TestParseVoteRevealed_Valid(t *testing.T) {
	event, err := domain.ParseVoteRevealed(validVoteRevealedJSON())
	if err != nil {
		t.Fatalf("parse: %v", err)
	}
	if event.ProposalID != "p1" || event.OptionID != "1" {
		t.Fatalf("unexpected payload: %+v", event)
	}
}

func TestParseVoteCommitted_Valid(t *testing.T) {
	raw := []byte(`{
		"event_id": "e1",
		"event_type": "VoteCommitted",
		"timestamp": 1,
		"source": "voting-contract",
		"version": 1,
		"proposal_id": "p1",
		"voter_pubkey": "v1",
		"commitment": "abc"
	}`)
	event, err := domain.ParseVoteCommitted(raw)
	if err != nil {
		t.Fatalf("parse: %v", err)
	}
	if event.Commitment != "abc" {
		t.Fatalf("got commitment %q", event.Commitment)
	}
}

func TestParseProposalCreated_ValidationErrors(t *testing.T) {
	_, err := domain.ParseProposalCreated([]byte(`{
		"event_id": "e1",
		"event_type": "ProposalCreated",
		"timestamp": 1,
		"source": "voting-contract",
		"version": 1,
		"proposal_id": "p1",
		"title": "T",
		"options": ["only"]
	}`))
	if !errors.Is(err, domain.ErrProposalCreatedNoOptions) {
		t.Fatalf("got %v, want %v", err, domain.ErrProposalCreatedNoOptions)
	}
}

func TestParseVoteRevealed_IgnoresUnknownFields(t *testing.T) {
	raw := []byte(`{
		"event_id": "e1",
		"event_type": "VoteRevealed",
		"timestamp": 1,
		"source": "voting-contract",
		"version": 1,
		"proposal_id": "p1",
		"option_id": "1",
		"voter_pubkey": "v1",
		"future_field": true
	}`)
	event, err := domain.ParseVoteRevealed(raw)
	if err != nil {
		t.Fatalf("parse: %v", err)
	}
	if event.ProposalID != "p1" {
		t.Fatal("expected proposal_id")
	}
}
